src/core/server/saved_objects/migrations/actions/wait_for_index_status.ts

```diff
@@ -0,0 +1,115 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import * as Either from 'fp-ts/lib/Either';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import { ElasticsearchClient } from '../../../elasticsearch';
+import {
+  catchRetryableEsClientErrors,
+  RetryableEsClientError,
+} from './catch_retryable_es_client_errors';
+import { DEFAULT_TIMEOUT } from './constants';
+
+/** @internal */
+export interface WaitForIndexStatusParams {
+  client: ElasticsearchClient;
+  index: string;
+  timeout?: number;
+  status: 'yellow' | 'green';
+}
+
+export interface IndexNotYellowTimeout {
+  type: 'index_not_yellow_timeout';
+  message: string;
+}
+
+export interface IndexNotGreenTimeout {
+  type: 'index_not_green_timeout';
+  message: string;
+}
+
+export function waitForIndexStatus({
+  client,
+  index,
+  timeout,
+  status,
+}: WaitForIndexStatusParams & { status: 'yellow' }): TaskEither.TaskEither<
+  RetryableEsClientError | IndexNotYellowTimeout,
+  {}
+>;
+
+export function waitForIndexStatus({
+  client,
+  index,
+  timeout,
+  status,
+}: WaitForIndexStatusParams & { status: 'green' }): TaskEither.TaskEither<
+  RetryableEsClientError | IndexNotGreenTimeout,
+  {}
+>;
+
+/**
+ * A yellow index status means the index's primary shard is allocated and the
+ * index is ready for searching/indexing documents, but ES wasn't able to
+ * allocate the replicas. When migrations proceed with a yellow index it means
+ * we don't have as much data-redundancy as we could have, but waiting for
+ * replicas would mean that v2 migrations fail where v1 migrations would have
+ * succeeded. It doesn't feel like it's Kibana's job to force users to keep
+ * their clusters green and even if it's green when we migrate it can turn
+ * yellow at any point in the future. So ultimately data-redundancy is up to
+ * users to maintain.
+ */
+export function waitForIndexStatus({
+  client,
+  index,
+  timeout = DEFAULT_TIMEOUT,
+  status,
+}: WaitForIndexStatusParams): TaskEither.TaskEither<
+  RetryableEsClientError | IndexNotYellowTimeout | IndexNotGreenTimeout,
+  {}
+> {
+  return () => {
+    return client.cluster
+      .health(
+        {
+          index,
+          wait_for_status: status,
+          timeout,
+        },
+        {
+          // Don't reject on status code 408 so that we can handle the timeout
+          // explicitly with a custom response type and provide more context in the error message
+          ignore: [408],
+          // The client defaults to a 30s timeout whereas we use a default wait_for_status timeout
+          // of 50s so we need to make sure the client doesn't timeout the connection before the
+          // ES server responds
+          requestTimeout: timeout + 5,
```

Comment:
I noticed some unexpected timeouts in the actions integrations tests and it turns out elasticsearch-js has a relatively short connection timeout of 30s. Looking for migration `TimeoutError` in our cloud logs showed a few clusters which ran into this problem during putMappings. All occurrences were in 7.x but it makes sense that we don't want to kill the connection before the timeout has run out.

In order to bump the timeout by 5 seconds I thought it's easier to just use numbers in the action APIs and then convert it to a string before we send the request to ES